### ISSUE TITLE
Adding LCD, Keypad, and SD card support

### ISSUE THREAD

**Issue by bdurbrow** [gnea/grbl-Mega#77]
I'm thinking of setting this up on a common 3D printer platform (Arduino Mega 2560 + RAMPS 1.4 + LCD w/ SD card reader), and I'd like to make it run standalone (like Marlin does).

So, I figure I'm going to need to write/port a SD card reader module, keypad module, and LCD menu system module. Not a big deal, I'm a programmer. :-)

However, I would like some input from the official developers as to where I should hook into grbl... without making a mess of things.

So... where should I inject the G-code I have read off of the SD card? Where should I drop a call to my LCD update routine; and where should I drop a call to my keypad polling routine?

I'm going to want to support a Manual Pulse Generator hand wheel also... where is the proper place to inject jogging data? Or should I take over the Arduino when in MPG jog mode, and then update the machine position global variables afterwords (and if so, which variables should I update?)

Also, is there an officially sanctioned way of directly setting (from a C function, not by injecting a G-code into the input datastream) feedrate override, spindle override, and coolant override parameters?

Any pointers you can provide (er, no pun intended) would be much appreciated. Thanks!

**Comment by terjeio** [comment]
I am using function pointers to achieve [a similar goal](https://github.com/terjeio/GRBL_MPG_DRO_BoosterPack). But perhaps not the kind of pointers you are after...? Another kind of pointer may be this [ESP32 port](https://github.com/bdring/Grbl_Esp32) as it has SD card support, and perhaps it is possible to use its serial comms over blueetooth for GCode injection?

**Comment by dangerousderek** [comment]
I’ve been scouring the web looking for exactly this. I’m no coder, but I love grbl and I just want a stand alone system using a ramps board for a laser cutter I built. Any progress you make on this front would be awesome and I would be happy to beta test. Thanks!!!

**Comment by bdurbrow** [comment]
> scouring the web looking for exactly this

So was I. I started trying to modify MK4Duo (a Marlin variant) to remove the 3D Printer specific components, but have determined that they are too deeply embedded into the firmware for that to be practical.

So I'm doing this instead. So far, I've got the device drivers (SD, Matrix Keypad, HD44780 LCD) ported from the Arduino codebase; and partially integrated into the Grbl codebase (there's quite a bit of incompatibility between the base Arduino core and Grbl, in particular in use of system resources like timers; so I've had to re-code quite a bit of the low-level Arduino support libraries to make it compatible... it wasn't a matter of drag-and-drop-and-compile).

Unless/until I'm told otherwise, I'm injecting the data from the SD card and jogging into protocol_main_loop(); and have put a call to my UITask() function into protocol_execute_realtime() and protocol_exec_rt_suspend().

**Comment by Developkings** [comment]
i believe using a ramps friendly premade solution like the lcd 2004 or ultra would be the best solution. i dont know about coding that much, but would it be easier to adapt marlins integration of this components t grbl instead ? if i`m not wrong it already as a basic cnc menu on it

**Comment by bdurbrow** [comment]
> but would it be easier to adapt marlins integration of this components t grbl instead

No, it's not. I already went down that road.

> ramps friendly premade solution like the lcd 2004 or ultra

Hardware wise, that's what I'm doing. I'm using the Reprap Discount Smart Controller with a RAMPS 1.4 board. I'm not making my own hardware for this project - the whole idea was to use off-the-shelf parts.

For performance reasons, I'm using the character LCD, and not the graphic LCD - I'm hoping to get nice, quick DRO updates during jogging.

**Comment by bdurbrow** [comment]
Status Update:
Drivers are working, but could stand some optimization (reviewing the drivers that Arduino ships, they are excessively abstracted in their construction; and use slow IO functions. I intend to fix this before posting any code).

SD support is integrated into protocol.c; but is as of tonight untested.
Keypad support is fully functional.
Rotary encoder (i.e, the knob on the LCD pcb) support is fully functional.
LCD support is functional.

User Interface code is about half done. The UI Page system infrastructure is functional; the menu system is itself functional but needs more menu items; and the SD directory access is working (however, at the moment, nothing happens when you select a file - that code needs to be written yet).

Still to do:
SD file opening when you select a file.
Jogging.
Probing.
Part Zero setting.
MDI.
"Conversational" operations for common simple tasks that don't warrant going to the extent of firing up a CAD/CAM program (drill a hole, bolt hole pattern, cut a slot, machine a face flat, etc).

Jogging, probing, and part zero setting basically just involves sending the appropriate command to the main grbl routine; in the same way that a PC would initiate that operation. The infrastructure is in place for this, it just needs to be "wired up" to the keypad & menu system.

MDI requires implementing a text editor to enter the MDI command line on.
Conversational requires a similar text editor, but with a "form" style layout.

**Comment by dangerousderek** [comment]
I can't believe how much progress your making! It seems like such a no brainer I am amazed that no one has done it already, in fact, I kind of figured it must be extremely difficult to do since it hadn't been done. You're the man. Keep us updated and as soon as your looking for some guinea pigs I call dibs! I am pretty active in a couple of other forums and groups having to do with DIY CNC stuff and those folks are going to freak out if you manage to pull this off!

**Comment by bdurbrow** [comment]
I have now achieved my primary objective: GRBL, on a mega2560 with a RAMPS 1.4 & RepRap Discount Smart Controller, is streaming g-code from a file stored on a SD card, selected via LCD, and performing updates to the display faster than the liquid-crystal media can physically respond. Oh, and it's doing it while maintaining responsiveness to serial commands, too...

There's still additional work to be done, but as of right now, I'm pretty pleased with the way it's working. 😎

**Comment by dangerousderek** [comment]
Are you going to fork it from this github? I need it so bad! You’re a super hero. Good job!

**Comment by bdurbrow** [comment]
> Are you going to fork it from this github?

Yes.

MDI is working, and the UI for jogging is halfway there. I need to go dig out a MPG handwheel and cook up a cable for it, but that's probably going to have to wait for tomorrow.

Anyone who wants to test this should probably make sure they have the requisite hardware now, since it often takes weeks to months for shipping...

I suggest:

- A suitable SD card. (Anything that works with other Arduino SD software/hardware - like a 3D printer - should work... I'm using SD drivers originally sourced from the Arduino codebase). Make sure you format it correctly, though, with the tool from the SD Association (https://www.sdcard.org/downloads/formatter_4/index.html). I've had mixed luck using the built-in formatting tools provided by the OS (both OS X and Windows) on SD cards... sometimes it works, sometimes it doesn't. Using the tool from sdcard.org seems to avoid this issue.
- A mega2560.
- A RAMPS 1.4 w/ stepper drivers.
- A RepRap Discount Smart Controller LCD interface w/ Smart Adapter (the one with the 20x4 character LCD, rotary encoder w/push button, extra push button and SD card slot; NOT the ones with a graphic LCD, this firmware doesn't - and won't - support those).
- A 4x4 matrix keypad (there are inexpensive membrane-style ones available on eBay, search for "Arduino keypad 4x4" and pick one - or 10 - that's cheap; and if you get the same one that I got, I've got a PDF file that can be printed out and glued over the keypad to show what the key functions are)
- A suitable MPG handwheel. (The ones I'm using are 100ppr, 60mm handwheels, and can be found on eBay by searching for "CNC MPG 5v 60mm"; at the moment they are under $20usd).
- Switches for Cycle Start and Feed Hold (I'm using some arcade-style push-buttons, again sourced from eBay) and 5K pots for feed rate override and spindle override. These components will require soldering, so make sure you've got a suitable soldering station (temperature controlled iron, "helping hands" to hold stuff, magnification, good solder and flux, etc).
- Suitable cabling to attach everything. (I ordered some long male-to-female 0.100in "DuPont" style hookup wires, and some 1x2, 1x3, 1x8, 2x4, and 2x5 "DuPont" style shrouds from eBay; also I had on hand some suitable ring terminals for hooking up the MPG handwheel).
- And, of course, your CNC machine itself. 😁

**Comment by dangerousderek** [comment]
I’ve got everything but the 4x4 pad and the mpg wheel! I’m stoked!

**Comment by bdurbrow** [comment]
Well... that's a bummer. I can't find the MPG handwheels I had here... I already had a few spares on order, but they're gonna take a while to get here. I guess I should order a few more spares now, also... (of course, you know that just as soon as the new ones show up, the old ones will magically appear, also...)

So I'm having to use the (much coarser resolution) encoder on the LCD for now...

🤨

**Comment by dangerousderek** [comment]
Are you building an enclosure to hold all this biz? If so can I get some pics? I have a feeling you’re doing exactly the sort of stuff I need to do with my rig...

**Comment by dangerousderek** [comment]
So.. I hate to nag, but any updates? Thanks a ton!

**Comment by bdurbrow** [comment]
Sorry about the delay... I've had to wait on parts arrival - the MPG handwheels I had are still AWOL; and I had to order replacements. However, the new ones have arrived and I've made quite a bit of progress:

MPG Jogging works now, and I've not been able to over-run the CPU with pulses (I'm using an interrupt to capture the input pulses; the polling loop can, at high speeds, get swamped).

The infrastructure is in place for the conversational operations; and drilling is in place. Other operations (facing, hole interpolation, etc) need implementing.

MDI works.
Part Zero setting works... but there's a feature I want to add (to take into account tool diameter and touch-plate thickness).
Probing should be working, but requires testing.

I'm hoping to have a pre-beta up on github soon... perhaps a week or two.

**Comment by dangerousderek** [comment]
You don't have to apologize! I am just super excited and I am having a hard time reigning it in. I look forward to testing the pre-beta! It sounds like you're going above and beyond the call of duty!

**Comment by bdurbrow** [comment]
Getting real close...

**Comment by manisaranbala** [comment]
https://sites.google.com/view/arrsoft/автономные-контролеры/автономнй-контроллер-для-
фрезера?authuser=0

**Comment by bdurbrow** [comment]
@manisaranbala 

Um... I don't read Russian, and google translate is making hash out of that.

Is that supposed to be some sort grbl based controller?

**Comment by manisaranbala** [comment]
http://www.cnc-club.ru/forum/viewtopic.php?f=41&t=20190

**Comment by manisaranbala** [comment]
https://sites.google.com/view/arrsoft

**Comment by bdurbrow** [comment]
OK, here's something to play with.

WARNING: This is pre-beta software. This has not been tested with motors attached yet (my machine is in pieces at the moment). Although it seems to be working on my bench when connected to my oscilloscope, it could crash your machine! So, be careful... and be ready to pull the plug or hit your emergency stop (whichever is more applicable to your situation).

Also, there are known issues with the conversational programming... so don't try that yet.

I have included some photos of the wiring required for the keypad and MPG handwheel in the "UI Support" folder (and yes, that PCB is kinda' crusty!). A PDF file of the keypad overlay is in there also (it has versions in it to support 4 and 5 axis machines, even though grbl itself only goes to 3 at the moment).

Note that unlike a PC, you don't hold down the 'Shift' key; it's a toggle. Push shift, _then_ push the other key.

Probing and setting part-zero are done from the Jog page.

Be aware that editing the tool table does put a cycle on the EEPROM of the atmega... so you probably want to set up your tools once and leave it alone. So does setting part-zero from the Jog page (I intend to implement a settings menu later on that will allow this to be disabled, so that you're not burning EEPROM cycles every time you run a part).

Documentation still needs to be written.

And yes, eventually this will get a proper github repository. I just wanted to get something up here for you all to play with.

😀

[grbl-Mega.zip](https://github.com/gnea/grbl-Mega/files/2605324/grbl-Mega.zip)

**Comment by dangerousderek** [comment]
I’m super excited to mess with this. With the holidays I’ve been to busy to experiment in the shop, but I’m hoping this week to give it a try. One quick question- I still don’t have an mpg wheel (on a boat from China) will the rotary encoder on the reprap discount full graphic lcd work? I’m super stoked to try all the awesome built in features you’re working on, but for now I would settle for being able to see the status of the machine and do basic moves on the axis’s. Thanks again, you’re the man.

**Comment by bdurbrow** [comment]
Yes, the MPG is optional equipment. Recommended, but optional.
Just disable (by commenting it out with a // )this line in config.h:
#define USE_UI_ENCODER_B

This will redirect the jogging inputs from the MPG inputs to the main UI knob.

Minimum hardware requirements for using the UI:
Arduino Mega 2560 (obviously... :-)
RAMPS 1.4
Reprap Discount Smart Controller 20x4 LCD & rotary encoder panel.
4x4 Matrix Keypad

**Comment by ghost** [comment]
Here's a solution for laser burners/engravers:
https://openbuilds.com/threads/arduino-based-off-line-cnc-controller-for-laser.13315/

**Comment by bdurbrow** [comment]
@dangerousderek - 

> on the reprap discount full graphic lcd work

Uh oh, I hope that's a typo... this does **_NOT_**, and will not, support **_GRAPHIC_** LCDs.
20x4 HD44780-compatible Character LCDs **_ONLY_**.

If you need to order one, and don't want to wait for the slow boat from China; Amazon has them with Prime Shipping for about $12USD... on eBay they can be had from China for under $8USD at the moment.

For example:
https://www.amazon.com/BIQU-Display-Controller-2004LCD-Adapter/dp/B01FLU9X5Q/

@RunnyBore604 -
It looks like that's not open source?

**Comment by dangerousderek** [comment]
@bdurbrow I think I’ve got one of those laying around. I was initially planning to use one of these:
https://www.amazon.com/KINGPRINT-Graphic-Display-Controller-Adapter/dp/B076WQQX5K/ref=mp_s_a_1_1?ie=UTF8&qid=1543270471&sr=8-1&pi=AC_SX236_SY340_QL65&keywords=reprap+full+graphics+display

But I failed to read the 20x4 detail in your BOM! Thanks!

**Comment by easytarget** [comment]
Sorry OT ; but not sure how best to address this..

> @RunnyBore604 -
> It looks like that's not open source?

He's cross posting/promoting from his own OpenBuilds project here, (check user details). Over there he says he wont share the source code.. Kinda against the ethos, and if he's also lurking here then I imagine GRBL is involved, GPL etc.. Anyone got a Openbuilds account and want to flag him?

**Comment by Bazza1959** [comment]
Love your work bdurbrow, as soon as I came across this page I immediately dug out the ramps and mega stuff i purchased over a year ago and got all excited and started uploading your work, but uh oh, when I reread everything I found that I have one of those graphics lcds, so now have to wait for slowboat, as I am in ozzieland, but not to worry too much as I can still be printing all the components for the cnc in the meantime, thank you very much for all your hard work, I really do appreciate it, Barry

**Comment by BillEng** [comment]
Thank you bdurbrow, just what I have been wanting.
I have tried this on hardware that has previously had Marlin on  it (MPCNC version), I have no stepper drivers, no keypad and no second encoder.
Connecting with bCNC I can run a gcode program, love the real time DRO, the LCD updates with feed and spindle overrides and preset Gcodes adjusted in bCNC but the push button does nothing, I can turn to different menu pages also SD card not recognized, tried two one from my 3D printer (8gig) and one I use on the MPCNC (16gig).
I have tried commenting out the second encoder as mentioned above, no change. 
Am I missing something just running without anything else attached to Ramps.
 Thanks again, I can wait.

**Comment by BillEng** [comment]
Well I feel a right plonka with my comments above I had read the post a 100 times but obviously not absorbed them, I think I'll go dig a hole and bury myself.
I have added a keypad and it is working to the point of development.
The encoder on the LCD is too sensitive, can this be tweaked.
 I could buy a second encoder that has twice the PPM of the one mentioned above at 400 PPM could this be accommodated.

**Comment by Bazza1959** [comment]
your not the only one, I am a plonker also, after purchasing the required components I have just realised you guys are setting up a lathe or a milling machine where as I am doing a cnc router is there anyway I can still setup the keypad and lcd screen for my setup and if so what do I need to do before I go dig that hole

**Comment by BillEng** [comment]
It's grand to know I'm not alone in the world Bazza1959.
I'm using a CNC router, as yet I can see that it would suit my needs, just ignore the features that are not needed or maybe (a big maybe in my case) mod the menu.
As mentioned above at the moment I am using Marlin, no disrespect to the devs but I hate all the 3D printer stuff in the menu.
I have a remote pendant that helps but can only jog in increments, my wish list would be.
Manually using the machine, handwheel.
Feed rate override while running a gcode.
Run gcode from SD card, semi headless.
To enter gcodes and/or some preset gcodes from a keypad.
The developmet so far seams to fit the bill.

**Comment by bdurbrow** [comment]
Sorry about the delay, I just haven't had much time to work on it lately. I should have something new up soon though...

@BillEng - The encoder on the LCD can be adjusted somewhat by changing it's configuration (UI_ENCODER_A_MODE in cpu_map.h); but it was designed for the encoder on a RepRap Discount Smart Controller; if you are using something else with more pulses per detent then I'll have to add a sub-divider (not a big deal; I just need to know to do it).

@Bazza1959 - I'm intending this to be usable on both mills and routers. If you can live without threading, a lathe could be setup also... but at the moment grbl has no support for syncing the axis movement to the spindle.

**Comment by BillEng** [comment]
bdurbrow, please don't retire you will have less time (-:, don't do any mods for an encoder at this time, I asked because I could buy one from my home country (UK) without the hassle of direct from China.
When I was trying the firmeware I did have an odd problem, the second row of buttons is a duplicate of the first row, checked through the files as best I could.
Flashed a test prog onto the Mega to test the keypad and my extension lead, all OK, I even tested the continuity of the Ramps tracks from AUX2 to where they connect to the Mega again all OK.
Thanks again.

**Comment by cadenote** [comment]
Great work @bdurbrow ! Thank you very much. It is what I was searching for.
When I have played with it, I have noticed that files coming from gcodetools ( inkscape )  stop when encounter a line with only a comment or a void line in the nc file.
To overcome this issue I have introduced minor changes in protocol.c and SdFile_grbl.cpp files.
I hope this will be useful and not to spoil anything.
[grbl_Mega_proposal.zip](https://github.com/gnea/grbl-Mega/files/2905655/grbl_Mega_proposal.zip)
x:)

**Comment by Bazza1959** [comment]
@cadenote is there anything needed to be done with the files if used on a cnc router or can I just simply replace the old files with the new ones T.I.A.
Barry

**Comment by cadenote** [comment]
Just replace them in libraries/grbl directory and upload again grblUpload.ino into your arduino.
:smiley:

**Comment by Bazza1959** [comment]
BillEng if using an arduino mega and a ramps 1.4 board do not connect power to the ramps board without having stepper motors connected, Th reason is that it will burn out the stepper drivers and maybe even the ramps board, ask me how I know this lol very frustrating, it took me a month to work out why my steppers wouldn't move and another month to get new drivers from china and now that they have arrived I can see if the boards are stuffed also, but I did order new ramps board and an uno cnc board as well and they should be here in a few days

**Comment by sinfocomp** [comment]
Hello, I'm trying but can't load any file, what is the correct file extension to use?
Also if I disable encoder I get compile error

**Comment by cadenote** [comment]
> Hello, I'm trying but can't load any file, what is the correct file extension to use?

I also had the same problem. I had to dig into the software to find that extension has to be .nc or .NC

**Comment by sinfocomp** [comment]
can you please let me know how to load a file, i used .nc and .NC but still my display shows NO FILE

**Comment by Bazza1959** [comment]
sinfocomp what file are you trying to load we are not sure of what you mean exactly. Are you trying to load cadenotes file or the original GRBL file from bdurdro or what?

**Comment by sinfocomp** [comment]
I have loaded firmware however
I have two problems
I cannot compile without the mpg
And I get NO FILE on lcd even if my SD card contains a testjob.nc file
Thanks

**Comment by Bazza1959** [comment]
you should be able to find and comment out the mpg wheel so that fixes your first problem but you will have to wait until bdurbro can answer the second one. I am not at the stage of needing this firmware as I am still waiting on a new ramps board and drivers as I stuffed my old ones

**Comment by sinfocomp** [comment]
I commented the mpg line as noted somewhere on this thread however I get compile errors unless I Un comment the mpg line

**Comment by cadenote** [comment]
@sinfocomp  in order to see your file in the display you have to presss menu key and after that you have to select SD Card option. Regards.

**Comment by sinfocomp** [comment]
Cadenote thanks, and in order to not use an mpg? Just one line needs to be commented on source?

**Comment by cadenote** [comment]
Well, up to I guess, to compile without errors you need to comment  out two lines in config.h : 
#define USE_UI_ENCODER_B
#define  USE_UI_ENCODER_B_ISR
However in that case jog mode has to be tested deeply.

**Comment by sinfocomp** [comment]
Hello, i need some help making this work on a MKS Base V1.5 board,
my problem so far is the keypad pins,can someone please helpme on remaping pins to A11,A12,D11,D12, and A13,A14,A15,D0,  Thanks in advance

**Comment by sinfocomp** [comment]
Got a ramps 1.4 board for testing and up and running, any information on where the menu text is, i would like to translate to spanish.
Thanks

**Comment by cadenote** [comment]
Mainly they are in UIMenus.cpp but  also there are lots of messages in report.c
Regards,

**Comment by sinfocomp** [comment]
> Hello, i need some help making this work on a MKS Base V1.5 board,
> my problem so far is the keypad pins,can someone please helpme on remaping pins to A11,A12,D11,D12, and A13,A14,A15,D0, Thanks in advance

any help on re asigning the keypad pins ?

**Comment by Bazza1959** [comment]
sinfocomp I feel you are asking the wrong people by being on this site, the title of this is GRBL-Mega.
the MEGA stands for an Arduino Mega 2560 board with a Ramps 1.4 cnc shield on top, and I am unsure that any of our programming would work on boards that are not the same. Maybe some of the other guys can correct me if I am wrong as I am not an expert at it, which is why I am on this site and I sincerely apologise if I am wrong

**Comment by Bazza1959** [comment]
OUCH $55 aussie just for the board, my mega and cnc shield was less than $4 delivered plus another few for the drivers. Sorry but I have no idea on how to remap pins
Footnote: still waiting for new ramps boards after blowing mine up

**Comment by Bazza1959** [comment]
OUCH $55 aussie just for the board, my mega and cnc shield was less than $4 delivered plus another few for the drivers. Sorry but I have no idea on how to remap pins
Footnote: still waiting for new ramps boards after blowing mine up

**Comment by sinfocomp** [comment]
finaly got it working !!!!
just dont seem to know where to hookup feed hold button ???
looked at cpu_map.h but feedhold is analog 9   and looks like is being used for keypad, anyone here knows where to conect feed hold ?

**Comment by sinfocomp** [comment]
please help, where do i connect the pot for feed overide ?

**Comment by lgojniczek** [comment]
Hi makers

I would like to ask how to clone the Y axis? Is it possible ?

I have cnc router that have two stepper and 2 motors on Y axis - one is reversed in rotating. My cnc is made of mega2560 and ramps 1.4 with lcd2004

Thank you for help

**Comment by zvizdic** [comment]
(Minimum hardware requirements for using the UI:
Arduino Mega 2560 (obviously... :-)
RAMPS 1.4
Reprap Discount Smart Controller 20x4 LCD & rotary encoder panel.
4x4 Matrix Keypad)
Can I use external drives ?

**Comment by JoeAndrew** [comment]
> > Are you going to fork it from this github?
> 
> Yes.
> 
> MDI is working, and the UI for jogging is halfway there. I need to go dig out a MPG handwheel and cook up a cable for it, but that's probably going to have to wait for tomorrow.
> 
> Anyone who wants to test this should probably make sure they have the requisite hardware now, since it often takes weeks to months for shipping...
> 
> I suggest:
> 
> * A suitable SD card. (Anything that works with other Arduino SD software/hardware - like a 3D printer - should work... I'm using SD drivers originally sourced from the Arduino codebase). Make sure you format it correctly, though, with the tool from the SD Association (https://www.sdcard.org/downloads/formatter_4/index.html). I've had mixed luck using the built-in formatting tools provided by the OS (both OS X and Windows) on SD cards... sometimes it works, sometimes it doesn't. Using the tool from sdcard.org seems to avoid this issue.
> * A mega2560.
> * A RAMPS 1.4 w/ stepper drivers.
> * A RepRap Discount Smart Controller LCD interface w/ Smart Adapter (the one with the 20x4 character LCD, rotary encoder w/push button, extra push button and SD card slot; NOT the ones with a graphic LCD, this firmware doesn't - and won't - support those).
> * A 4x4 matrix keypad (there are inexpensive membrane-style ones available on eBay, search for "Arduino keypad 4x4" and pick one - or 10 - that's cheap; and if you get the same one that I got, I've got a PDF file that can be printed out and glued over the keypad to show what the key functions are)
> * A suitable MPG handwheel. (The ones I'm using are 100ppr, 60mm handwheels, and can be found on eBay by searching for "CNC MPG 5v 60mm"; at the moment they are under $20usd).
> * Switches for Cycle Start and Feed Hold (I'm using some arcade-style push-buttons, again sourced from eBay) and 5K pots for feed rate override and spindle override. These components will require soldering, so make sure you've got a suitable soldering station (temperature controlled iron, "helping hands" to hold stuff, magnification, good solder and flux, etc).
> * Suitable cabling to attach everything. (I ordered some long male-to-female 0.100in "DuPont" style hookup wires, and some 1x2, 1x3, 1x8, 2x4, and 2x5 "DuPont" style shrouds from eBay; also I had on hand some suitable ring terminals for hooking up the MPG handwheel).
> * And, of course, your CNC machine itself. grin

I've tried the code and it works, without steppers.
Is there a wiring diagram of where to wire exactly? I think about the buttons the most.
In the meantime, I try to merge your projects with the lathe sync code.

**Comment by Saur0o0n** [comment]
Hi,
Any chance for straight Makefile for this project? (I assume that people who have compiled it used bbproject file to do it).
Any ETA when this project can drop on git? (as promised :) )

**Comment by erosenst** [comment]
Hello, It needed redefine boolean to compile and download to board.
I don't have mpg wheel so commented out USE_UI_ENCODER_B &ISR.
However menu dosn't open. It only changes three DRO-menus when rotating encoder. Encoder button not working or what code pices need change?
 Where should keypad to connect. I have green 4x5 matrix keyboard. AUX2 and one pin from AUX1??
https://www.aliexpress.com/item/1pcs-4x5-Matrix-Array-20-Key-Membrane-Switch-Keypad-Keyboard-Control-Panel-Microprocessor-Keyboard-Controller-for/32799058868.html
Connection diagram??

Any ideas how combine two motors Y to second motor E0 or E2? And how loose Y carriers? Is there code to home with two motors against two home switch? (motors need to be separated during homing, so only hard coding bitmasks is not good idea (?))

And plan is to use that code for plasma cutter table. Ideas how add THC (torch heigth control). There is TEMP0, TEMP1, TEMP-BED in ramps 1.4 so some inductor + resistor network may work to inject suitable volt range from torch to some temp input (?)

**Comment by zvizdic** [comment]
Complete all components and everything works great 
Big thanks to 
![3082A817-0ACE-411D-ACFA-D09ED9828C1C](https://user-images.githubusercontent.com/50154737/60307515-f9262b00-9912-11e9-948e-9996652f6d72.jpeg)
bdurbrow

**Comment by Saur0o0n** [comment]
Once again - how to compile this? Makefile is unchanged from GRBL Mega and is missing files. Can't say if there is anything important is in bbproject - since bbedit is closed source app only for mac.

**Comment by Bazza1959** [comment]
zvizdic would you be willing to share the gcode or stl for the board you 3d printed

**Comment by cadenote** [comment]
> Once again - how to compile this?

One way to do that is adding grbl directory as a library and after that, flashing grblupload.ino to your board.
You can follow step 2 of that link: https://www.instructables.com/id/How-to-Installuse-GRBL-With-Your-Cnc-Machine/

**Comment by Saur0o0n** [comment]
> One way to do that is adding grbl directory as a library and after that, flashing grblupload.ino to your board.

Thx. This way it works (unless compiling on Arduino IDE - haven't uploaded it yet) - I've used to compile grbl with avr-gcc and grbl Makefile. And this way (even after some modification to add missing objects) - it still had some unresolved dependencies.

**Comment by Bazza1959** [comment]
correct me if I am wrong but bdurbrow wrote this for loading onto an arduino mega 2560 so you have to expect some problems using non Arduino boards and non  Arduino IDE to upload to that board

**Comment by zvizdic** [comment]
> zvizdic would you be willing to share the gcode or stl for the board you 3d printed

I’m using off shelf box 20$ cad .

**Comment by zvizdic** [comment]
I am using non arduino mega no problem uploading

**Comment by coilheaddesigns** [comment]
This seems like it should be getting way more attention than it is.
Getting ready available hardware for marlin compatible with grbl, 
As grbl is a lot more user friendly for cnc machines.

**Comment by coilheaddesigns** [comment]
Anyone got this working?
Mind sharing some pictures. Please.

**Comment by Saur0o0n** [comment]
> Anyone got this working?
> Mind sharing some pictures. Please.

You have some pictures in zip file ("UI Support" folder). I'm kind of finishing my second machine now (one more month probably) - I'll definitely share what I've done, but I haven't decide yet if I'll use this code. So far it runs on plain grblmega.

**Comment by coilheaddesigns** [comment]
Mind sharing some pics.
Just got
Ramps 1.4 with lcd. And mega
If grbl works great without lcd I don't mind not having the lcd.
Looking to use it for mill and perhaps a rotary attachment.

**Comment by coilheaddesigns** [comment]
Hi Guys so got this working beautifully,

how do i use Extruder port for additional Y axis motor?

my cnc running dual 1.2A Nema 17 motors.

**Comment by coilheaddesigns** [comment]
Is there a way to add the lcd display from this version to grbl-megaX5

Can someone please advise.

**Comment by zvizdic** [comment]
I work in Cnc programming and machining day in day out .I think as a professional in the field this is best hobby cnc setup you can have .
Thanks again

**Comment by Bazza1959** [comment]
so zvizdic I take that as a no that you won't share the code, I am a disabled old man and you can't be bothered to help a fellow out, yet here you are using bdurbro's hard work to make it easy for you, yep lots of wankers around

**Comment by zvizdic** [comment]
> so zvizdic I take that as a no that you won't share the code, I am a disabled old man and you can't be bothered to help a fellow out, yet here you are using bdurbro's hard work to make it easy for you, yep lots of wankers around

Store bought item  I didn’t 3D print the box .
HAMMOND MANUFACTURING 1598ABK ABS Plastic V0 Aluminum Instrument Enclosure BLACK

**Comment by Bazza1959** [comment]
zvizdic oh sh!t will you please accept my very sincere apologies, I am truly sorry, I can only put it down to how bad my pain was yesterday and I was a tad grumpy. Again I am sorry for the outburst

**Comment by coilheaddesigns** [comment]
@bdurbrow how to clone Y axis on your version please

**Comment by spacedeep** [comment]
Доброго всем, можно ли на данной прошивке использовать reprap full graphik discount smart controller 12864? И подскажите как это сделать. Спасибо

**Comment by bdurbrow** [comment]
You ever have one of those ~~days~~ ~~months~~ years? Eee-gads I've been busy... and not on this, alas...

Attached is an archive of my development folder. It compiles and runs on my bench, but there are NO guarantees about it's stability.
[grbl-Mega.zip](https://github.com/gnea/grbl-Mega/files/3388571/grbl-Mega.zip)

**Comment by bdurbrow** [comment]
**Re: editing and compiling:**
Just to confirm - yes, the .bbprojectd file is just for using BBEdit on a Mac to edit the source code. It's not a makefile, and it doesn't load into the Arduino IDE or anything like that. If you are not using BBEdit you may safely ignore it.

To compile this, use the standard procedure for compiling grbl with the Arduino IDE. It's a direct modification of the original grbl-mega repository; I've only added to it, not changed the fundamentals of how it's built. I've tried to keep my changes limited to adding files and a few lines of code in certain spots to hook into the system; such that it's easy to port to future versions of grbl.

**Re: hardware hookup.**
I'm afraid that I only have what you find in the .zip archive (check the UI Support folder) for data on how to hook it up; although in principal it's portable to other non-RAMPS setups; I've not tried that. I don't think I'll ever have comprehensive wiring diagrams just due to the sheer number of possibilities... but I would like to work up something better than I have now. I know it won't anytime soon, though... (right now I'm at 100% time saturation on other non-optional projects).

If you aren't using my exact setup (RAMPS, etc), I would suggest getting an oscilloscope to help when figuring out what pin is mapped where... even a cheap ebay/amazon/bang-good/tao-bao special will work for this - or even just a logic probe. The idea being that say, if you don't know where the X step pin is showing up on your board's physical pinout, you command a long X move (with no other axes active), and probe all the pins until you find one that's toggling up and down in the kHz range.

**Re: axis cloning**
A quick look just now indicates that this version of grbl isn't going to be able to do software axis cloning easily. It makes assumptions that there are exactly one step output bit per axis and exactly one direction output bit per axis.

So, I would suggest either a: running both motors off of a single driver chip (not something I'm very fond of myself) or b: running the motors of their own driver chip, but connecting both driver chips to the same physical output on the breakout board (if that's a RAMPS board, you might have to get some "dupont" patch wires and hook into the appropriate spots on the stepstick sockets; and if you wish to stick to stepstick compatible driver boards, an ebay search for "stepper driver carrier board" turned up some possibilities for hooking up stepsticks external to the RAMPS board).

**Re: @zvizdic's enclosure**
Nice! :-)

**Comment by qulet** [comment]
Bdurbrow thank you very much for sharing the code!!!!!!!!!
I assembled this control. Jog works on all axes.
I try to run the g-code from the SD card, but I can not ...
  After selecting the file, the file name and X.Y.Z coordinates appear on the LCD
Pressing enter, encoder LCD does not work. Where the Cycle start and Feed Hold buttons should be connected
How does the g-code file run?
Please help

**Comment by qulet** [comment]
![IMG_8452](https://user-images.githubusercontent.com/30979927/61232603-0030a400-a72f-11e9-9666-379b7068b0ce.jpg)
![IMG_8453](https://user-images.githubusercontent.com/30979927/61232619-06268500-a72f-11e9-8da9-204c03acd6d1.jpg)
How can this G-code file be fired????

**Comment by qulet** [comment]
MDI and JOG Working I can not check homme ...
G-code is fired after pressing enter :)))) Must be clean without comments.

I am asking for Pin out Ramps 1.4 for this GRBL build  
 Home X,Y,Z  Pin.....
Cykle start  Pin.....
Feed hold  Pin.....
Pots 5k feed rate  Pin......
Spindle PWM Pin....
Spindle Enable Pin.....
Probe Pin.....


I tried :
Reset / Abort A8
Feed Hold A9
Cycle start A10
  I attached pins with GND
Unfortunately, this did not work ...: ((((

**Comment by qulet** [comment]
I know that the answer to my question is in the file cpu_map.h ... but I can not properly interpret it.
  Can anyone do this for me?
Bdurbrow - it's probably busy, I see after the comments that several people are testing this compilation. Please help and answer about the missing PIN_in and PIN_out mentioned above.
I apologize for my English - it's a Google translator.

**Comment by bdurbrow** [comment]
OK, the good news is one of the things that was running me ragged is now off my plate. Only fifteen-bazzillion more to go!

@qulet - 

Homing has to be enabled by parameter $22. You can do this in the Arduino IDE by opening the serial monitor, and sending a $22=1 command to grbl.

All the "Home All Axes" command found in the System menu does is check that everything is OK and then issue a "$H" command to the main grbl interpreter. Look on line 117 of UIMenus.cpp to see what I mean.

This image might be of help in figuring out the pin mapping (which as you mentioned is in cpu_map.h; for RAMPS the active configuration starts on line 138).
[https://reprap.org/wiki/File:Arduinomega1-4connectors.png](https://reprap.org/wiki/File:Arduinomega1-4connectors.png)

Look at line 224 of cpu_map.h for limit switch configuration.
Other pins can be found starting on line 299.
Override potentiometer inputs are starting on line 664 of config.h.

While investigating the comment handling situation (it's *supposed to* filter comments out), I discovered a nasty little memory leak in the SD library... opening a file works the first three times... and then crashes.

Wait - how can it be leaking memory if it never... calls... malloc?

That's right ladies and gentlemen... this SD library for an Arduino calls malloc in it's constructor, and it has *no* destructor! So every time a File is allocated on the stack, it allocates a SdFile on the heap, and never frees the memory.

So, to fix this I need to rewrite the File class to directly embed the SdFile and not refer to it by a pointer to malloc'd memory. Yes, I could just add a destructor that frees the pointer, but that still leaves the issue that Arduino's don't have much memory to begin with and due to heap fragmentation, using malloc is not that great of an idea.

I have no idea how long this is going to take me to do... it depends a lot on if anybody drops anything in my lap.

**Comment by cadenote** [comment]
> G-code is fired after pressing enter :)))) Must be clean without comments.

Maybe you can avoid that as stated above in https://github.com/gnea/grbl-Mega/issues/77#issuecomment-467439164  :smiley: 
Regards,

**Comment by zvizdic** [comment]
I am rebuilding Tor Cam Cnc lathe to be controlled with this controller and GrblGru software .
Everything works so far good other then ability to feed  hold and resume witch I am planing to achieve by soldering directly to pins 9 and 10 of Arduino .
Any suggestion?

**Comment by zvizdic** [comment]
Reset or stop is not working on RepRap Discount Smart Controller .

**Comment by zvizdic** [comment]
![CE370297-8B23-43B4-B5B7-A4E0250DAEB5](https://user-images.githubusercontent.com/50154737/61583542-82daaa00-ab07-11e9-829a-55ea9661974e.jpeg)
![1EA64FCF-6234-487F-9474-83FB395C8ED5](https://user-images.githubusercontent.com/50154737/61583543-82daaa00-ab07-11e9-9a7e-515dac18caf2.jpeg)

**Comment by bdurbrow** [comment]
I think I have the SD library fixed... and if I have this right (I make NO claims to be a git expert) I should be able to upload a proper fork to github as soon as I get home (I’m sitting in a restaurant right now posting from my iPhone).

@zvizdic - that button probably is working on your hardware, but it’s not doing what you expect (understandable considering the only documentation is comments in the source code). It’s used by the UI system as a “back” or “cancel” button.

If you’re looking for an actual reset button, there is one on the RAMPS board, and I think there might be pads for one on the LCD pcb, but I’m not in front of my bench so I’m not sure...

**Comment by bdurbrow** [comment]
Um... I _think_ that worked.

This is the new, proper, github fork for this project:

https://github.com/bdurbrow/grbl-Mega

Additional updates will happen there. Likewise, any bugs found should be reported on that fork.

Items of note:

- The SD library should now be working properly (I've removed the calls to malloc, everything is now either a global or on the stack).
- The comment & empty line filtering should be working correctly.

However, this remains pre-beta; and there is no guarantee of proper function. Actually, I'd be rather surprised if it _was_ bug-free...

**Comment by zvizdic** [comment]
What about feed hold and resume am I going to be ok to hack into Arduino pins 9 and 10

**Comment by bdurbrow** [comment]
@zvizdic  -
You shouldn't need to; look at line 305 of cpu_map.h.

    #define CONTROL_RESET_BIT         2  // Pin D47 - RAMPS Aux 4 Port
    #define CONTROL_FEED_HOLD_BIT     4  // Pin D45 - RAMPS Aux 4 Port
    #define CONTROL_SAFETY_DOOR_BIT   6  // Pin D43 - RAMPS Aux 4 Port

    #define CONTROL_CYCLE_START_BIT       5 // Pin D32 - RAMPS Aux 4 Port

**Comment by qulet** [comment]
bdurbrow
Thank you very much for the Pin In and Pin Out hints. I understand that you need to solder the cables to the Aux 4 ramp. Because there is a PCB connector from the LCD.

**Comment by qulet** [comment]
@zvizdic
Congratulations on the execution of a beautiful lathe design.

**Comment by zvizdic** [comment]
Thanks for pi out bdurbow !
The lathe is a 1990is school model for students to learn Cnc programming and it was controlled by a Pc true printer port .

**Comment by bdurbrow** [comment]
I think I'd probably solder to the adapter board, rather than the RAMPS itself...

I have a couple of large CNC mills here that need new electronics... but these machines are too big for a stepstick (one is a Sieg X3 bench mill, the other is a Shizouka AN-S knee mill that's over 7 feet tall... er... over 2 meters). I am intending to cook up a board that takes the same pin assignments as the RAMPS + Smart Controller combo; but just breaks out the stepper driver step & direction pins and directly has connectors for the LCD/SD/encoder PCB, keypad, jog wheel (MPG) and other switches & pots. When I will get this done... I have **_no_** idea... but when it *does* actually happen, and I've tested the board layout, I will post the gerbers so that other people can manufacture the boards should the RAMPS itself be insufficient for their needs (why not the original design files? because I use an oddball proprietary - and very much pre-alpha - PCB design program... nobody else has a copy... so the files wouldn't be useful).

**Comment by terjeio** [comment]
@bdurbrow : Another option is to make some Polulu-style "dummy" drivers and use drivers like DQ542MA with the RAMPS board- I [did that](https://github.com/terjeio/CNC_Boosterpack) for my lathe - scroll down to _Bonus design: Polulu style motor driver "dummy"_ to have a look.

**Comment by zvizdic** [comment]
I did not use dummy drivers ramps have marks for step and directions pins so you can tap in to them for your driver connection.

**Comment by qulet** [comment]
I uploaded a new improved grbl-Mega-edge code. I fired from the G-gode SD card generated in the Vectric Aspire program for the GRBL postprocessor. Unfortunately, after about 5 minutes of work, the engines stopped spinning, sending for unknown reasons was discontinued ...

It was a simulation of work as in the picture above only the motors of the x, y, z, axis, with no spindle.

The same G-code from the SD card fired from the computer on the GRBL-Panel sent via the USB cable after com8, stopped at 6876 lines ....


[V-Carve 2.zip](https://github.com/gnea/grbl-Mega/files/3418772/V-Carve.2.zip)

**Comment by qulet** [comment]
On my Mach3 machine, I made this calendar.
https://drive.google.com/file/d/1-6xKRXwPXY7beysxDcEU91E6Z2fnPKyw/view?usp=sharing
https://drive.google.com/open?id=1bcXwgqRb9gMRJPkNCTmrvFrbDMGSQhC8

**Comment by bdurbrow** [comment]
@qulet - moving forwards, we should be posting issues to the new github repo.

I have created an issue for this; and will investigate it when I get some time later tonight.

https://github.com/bdurbrow/grbl-Mega/issues/1

**Comment by bdurbrow** [comment]
> I think I'd probably solder to the adapter board, rather than the RAMPS itself...

After having spent the last several hours hunched over a microscope with a hot soldering iron in my hand, I think I'm going to change my mind on this.

If you have the equipment, you could de-solder and remove the relevant pins from the RAMPS card itself, leaving the pads open for you to insert wires into (this is what I ended up doing for a couple of RAMPS cards).

Otherwise, I might try carefully separating the black plastic that holds the header pins together on the pins in question (perhaps with a razor or x-acto style knife - using 'nippy' style flush cutters resulted in bending the surrounding pins); removing it; and then using needle-nose pliers to bend the header pins 90 degrees outwards so that you can hook 'dupont' style jumper wires to them. I have several more RAMPS cards here that I have not de-soldered yet, and I will give this a shot on them when I get a chance to do so (worst case; I have to de-solder the pins anyway, so...).

:smile:

**Comment by zvizdic** [comment]
> @zvizdic -
> You shouldn't need to; look at line 305 of cpu_map.h.
> 
> ```
> #define CONTROL_RESET_BIT         2  // Pin D47 - RAMPS Aux 4 Port
> #define CONTROL_FEED_HOLD_BIT     4  // Pin D45 - RAMPS Aux 4 Port
> #define CONTROL_SAFETY_DOOR_BIT   6  // Pin D43 - RAMPS Aux 4 Port
> 
> #define CONTROL_CYCLE_START_BIT       5 // Pin D32 - RAMPS Aux 4 Port
> ```

I tried using a safety door pin as my hold/resume since I have one on/off switch i place (see pics)but it’s not doing anything . Also try the feed hold and it works.
I would rather have existing on/off switch doo hold and resume 
Any chance for some help
Switch should be connected to control ground and pinD43 if I am right

**Comment by bdurbrow** [comment]
_**WARNING!!!**_

I just spotted an error in the Wiring Overview that I had up on the wiki... the 5v and ground pins were reversed on the keypad connector (RAMPS Aux 2).

I've corrected the Wiring Overview image and uploaded it - but if anyone has downloaded the WiringOverview.png file; you should delete the old one and download a fresh copy.

Fortunately I don't think it's caused any issues, because the 4x4 matrix keypad is a completely passive device; it doesn't hook up to the power lines at all... but... better safe than sorry!

:flushed:

**Comment by qulet** [comment]
I put it in the housing from the PC power supply. The first tests on my CNC milling machine indicate that all functions work.
 I have 3A engines so I cut out the sepstiki (they give too low current) and I will try to connect with the LPT cable to the Mach3 controller box where I have 4x TB6600. I will inform about the results.
Greetings to all CNC enthusiasts !!!
![IMG_20190816_155617](https://user-images.githubusercontent.com/30979927/63174277-e80bb780-c041-11e9-8e6a-b547dcf58945.jpg)


![IMG_20190816_155738](https://user-images.githubusercontent.com/30979927/63173898-2359b680-c041-11e9-8e0b-96407a940bc1.jpg)

**Comment by qulet** [comment]
Bdurbrow
Thank you again for sharing this GRBL-MEGA branch. The hardware and software solution is brilliant. My version is removed stepstiki. All out and in signals go to the LPT plugin. After cable on Breakout Board Mach3 and TB6600. After an hour of uploading the CNC program, I didn't notice any errors. I am very happy with the new possibilities of laser engraving.
Greetings

https://drive.google.com/file/d/1-JT9JfJ-psZ6LP40paNfMkOpIfdochT5/view?usp=sharing

**Comment by lloyd1941** [comment]
Gyday Bdurbrow
I have read your progress for the past year on this site, and like the result you have created, for something was said not possible. i started some 3 yrs ago to build a cnc router, i now have 2 + a lazer 
**my original cnc was to be **marlin on a ramps board**, stand alone,** never got to do this as was told to use the motion controller of Estlcam, that has worked so far. but it brain washed Marin??
I have 3 cnc machines all at the moment using GRBL, no complaints there except they need a PC to run, hence your development !! for stand alone Ramps Board (have 3 of them too). that i can use if this works?

**I ask do you have detailed instructions as to the installation of your program**

Thanks Mate
Lloyd

**Comment by lloyd1941** [comment]
oppps i for got to ask where do we ( IN the WORLD) FIND THIS. INFO
LLOYD

**Comment by bdurbrow** [comment]
Gah... do something else for a couple of days and somebody asks a question over here! 😃  Sorry about the delay...

OK, so... the github repository is here:

https://github.com/bdurbrow/grbl-Mega/

I few days ago I pushed to github the October 2019 update; which brings (amongst other things) an easier setup process: my version now compiles like any other Arduino sketch. No more messing with your library folder!

Basically, it goes like this:

* Make sure you have the latest Arduino IDE installed on your computer.
* Download & decompress the .zip of the repository from the github repository I linked to above.
* Make any configuration changes needed for your specific hardware to the config.h and cpu_map.h files. This will mostly be turning things on; like the MPG handwheel or rate override potentiometers.
* Open the .ino file at /grbl-Mega/grbl/grbl.ino with the Arduino IDE.
* Plug in your Arduino Mega 2560 to your computer.
* Select the port for your Arduino Mega 2560 from the Tools->Port menu.
* Click the Upload button in the Arduino IDE.
* If everything went well, your Arduino Mega 2560 should now be a CNC controller. 😄 

However; in-depth documentation still needs to be written. There is some information on the home page of the repository (which you _will_ need to read - configuration is a little different than the standard version of Grbl); and reading thru both the config.h and cpu_map.h files is also helpful. If you have any specific questions (or find any bugs!), **please** ask them on the Issues page of my repository. I usually try to check this page every day, sometimes more if there's a specific question/issue that I'm trying to help someone with.

https://github.com/bdurbrow/grbl-Mega/issues

**Comment by Elpablito** [comment]
Hi bdurbrow
At first congratulation for the fantastic job you have done, I am following you on github since last year when I start to think to make CNC. Sorry for my English I shall try to make my best.                                                                 I made a machine, bat it is my first one, so I am not an expert in working with such a kind of machine. The CNC is  made with some plastic parts good for wood or soft materials.  Actually, I have used an Arduino Uno with its shield for stepper drivers and it seems it is working properly, at least for the tests I have done.     For sure, since the beginning the idea was to have a standalone CNC so I am trying to replace Uno with a Mega that I was having somewhere. Unhappily, I faced these kind of problems that I am not able to solve.
1)	I have modified the hardware setting $0……$132 with UGS and when saved. If I disconnect the board and then reconnect the setting are the old ones not the new ones. It seems that save action do not write EEprom. 
2)	Why in the home cycle homing Z is optional and why is the third action, moving before Y and X and then Z?  If any reason I shall prefer to move Z as first  action.
3)	Which is the use of RAPID_OVERDRIVE potentiometer?
4)	The FEED_OVERDRIVE Potentiometer is to decrease the set speed of FEED?
5)	Is the SAFETY_DOOR the door of the cabinet you need to close before run or it is an escape way? In such a case what  is it doing?
Thanks

**Comment by bdurbrow** [comment]
@Elpablito - please use the issues section of my main repository. I'm trying not to overburden this official issues section with Q&A that really only applies to what I've been doing.  😄 

I have created a new issue over there, and answered your questions:

https://github.com/bdurbrow/grbl-Mega/issues/20

**Comment by bluffchuffer** [comment]
I have just managed to get your app running. I was looking for a better precision readout for the X and Y axis, something Marlin does not have on its current version status screen. I managed to upload to my mega/ramps CNC system  (runs a mill) though with a RSD Graphics Controller LCD screen. I removed the Graphics LCD and put the 2004 in. No joy, but I have struggled to make it function and guess what I found?
The cable adapter plug to fit to the ramps board are different between the two RepRap Lcd's! 
Something to lookout for. Still looking forward to testing it out.
One thing I also found, was the 'pendant' I used on this mega/ramps also works with this GRBL firmware - did not expect that - plenty of scope there I think - I try to get away from all the click button when i need simple commands.
Thanks for your efforts.

**Comment by how-eee** [comment]
Hi Bdurbrow!
I like your project so much. 
But I want to implement G33 spindle sync threading for my lathe. Its already done in this fork https://github.com/MetalWorkerTools/grbl-L-Mega
What do you think? Is it still enough performance of mega 2560 available in your fork? Will all that stuff work fast enough?

**Comment by austrian1** [comment]
hi Bdurbrow. 
great job. i have a question for you . where i connect the poteniometer and the push buttons. i cannot figure out

**Comment by bluffchuffer** [comment]
Hi, There is a Wiring Overview file in these pages somewhere that shows the Ramps connection clearly - seek and ye will find.

**Comment by Saur0o0n** [comment]
You have also [here](https://github.com/Saur0o0n/GRBL_Mega-pinout) all the pinouts described.

**Comment by austrian1** [comment]
yes thank guys i have already done it.

**Comment by how-eee** [comment]
Just for your information:
I had implemented G33 threading
Fixed couple errors in code
Tested on real lathe

See my repo https://github.com/how-eee/grbl_lathe_UI_threading

**Comment by lgojniczek** [comment]
> Just for your information:
> I had implemented G33 threading
> Fixed couple errors in code
> Tested on real lathe
> 
> See my repo https://github.com/how-eee/grbl_lathe_UI_threading

Hi
Is your fixed code can be used for cnc router? Or it is only for lathe?

**Comment by qulet** [comment]
Hello !
I made this design for 3 axis, it works fine. Now it needs a rotary axis for a tangential knife. Is the code author or someone else could add support for the rotary axis. There is a great GRBL-PLOTTER controller that generates g-code for the tangential knife. Now I am testing a GRBL-HAL which works from a PC without an LCD and SD card.

**Comment by erosenst** [comment]
https://github.com/bdurbrow/grbl-Mega/issues/25 
grbl-5xlt should do job for 4th (or 5th axis). It is based grbl 5X but not updated while. Be free to update if there is any in grbl 5X..

Author have been silence from 2019..

**Comment by qulet** [comment]
I understand that you are not planning to upgrade to 4th axis of your software.
Greetings !

**Comment by erosenst** [comment]
grbl-5xlt is 5-axis code. It is bdurbrow LCD pack + grbl-mega-5X. You find it link I give previous message. I run it with 4-axis plasma table and 5-axis grinding machine. Works. With SD-Wlan card code updates can done remotely with nautilus.

Mr.(?) bdurbrow disappeared from here 2019.. Current "best" what you can get is grbl-5xlt. (I did, but not updated while)

**Comment by kangah58** [comment]
salut les pros  quelqu un peux m'aider sur un controleur

**Comment by kangah58** [comment]
![Screenshot_20220616_142436_com chrome beta](https://user-images.githubusercontent.com/105965366/177489814-989cec13-c7ca-408c-a9d7-0ffec624a8ea.jpg)

**Comment by bdurbrow** [comment]
Sorry about disappearing - the past several years have been absolutely insane - one "dumpster fire" (to make metaphor) after another. Just to name a few of the more dramatic ones; I've been hospitalized a couple of times (luckily, the conditions were all curable - but they don't just send you home with an IV pump, apparently 😉 ), I've had _very_ time consuming construction problems with our house, and my car got totaled thanks to a hapless deer that jumped out of the bushes _right_ in front of me - ya know, the sort of stuff you'd expect to see on a soap opera. Until now, I haven't had *any* time to work on this; and my equipment has been un-accessable (too much stuff piled on-top/in-front of it, due to the construction).

I'm finally starting to dig my way out of it though; and do hope to resume development on this.

Before everything went sideways in my life I had some stuff written, but untested; once I've gotten my gear accessible and can test that I'll release it.

**Comment by kangah58** [comment]
salutEnvoyé depuis mon téléphone-------- Message original --------De : Britt ***@***.***>Date : dim. 25 sept. 2022 à 19:49À : gnea/grbl-Mega ***@***.***>Cc : kangah58 ***@***.***>, Comment ***@***.***>Objet : Re: [gnea/grbl-Mega] Adding LCD, Keypad, and SD card support (#77)

Sorry about disappearing - the past several years have been absolutely insane - one "dumpster fire" (to make metaphor) after another. Just to name a few of the more dramatic ones; I've been hospitalized a couple of times (luckily, the conditions were all curable - but they don't just send you home with an IV pump, apparently 😉 ), I've had very time consuming construction problems with our house, and my car got totaled thanks to a hapless deer that jumped out of the bushes right in front of me - ya know, the sort of stuff you'd expect to see on a soap opera. I until now, I haven't had any time to work on this; and my equipment has been un-accessable (too much stuff piled on-top/in-front of it, due to the construction).

I'm finally starting to dig my way out of it though; and do hope to resume development on this.

Before everything went sideways in my life I had some stuff written, but untested; once I've gotten my gear accessible and can test that I'll release it.

**Comment by Elpablito** [comment]
Hi bdurbrow
 I am very happy to hear of you again. Keep on going this is the life. I hope to hear that you are again fighting with a CNC.Best wishes
 Paollo

**Comment by NapierDeltic** [comment]
Hi bdurbrow! 
Great to see you  again. I'm sorry to hear bad happenings you faced but glad to see you are back. Hope you'll deal with them 
I do very much appreciate your concept of CNC controller. So many people have so many preferences... But at least from me this controller - soft/hard -  offers best balance of features for my taste, keeping also a clean, simple and reasonable cheap configuration. My believe is a lot of of people coming from mechanical pool of CNC beginners look for a simple but feature rich device with less programming hassle (somehow like Marlin/Ramps is for 3D printers).
So here is my take for your controller and I have made 2 of them (for 2 locations....) : 
![GRBL Controller 1](https://github.com/gnea/grbl-Mega/assets/81844337/31704a2b-2db9-4e00-80c4-83176cddf196)
Still having some (same) issue on both with correct running but I will address them in a ticket,

I know it is time consuming but I would like some tutorials; especially for menu items - being so complex. Maybe even videos.
For the rest, I presume Arduino Mega will upgrade soon to 32 bit and more memory, so less limitations. Also, in that case, a bit larger display would be helpful, but please keep everything as simple (or modular) as can be.
For myself  I'll keep mine in the current status, unless a simple drop in place software/hardware will appear.

**Comment by lgojniczek** [comment]
> Sorry about disappearing - the past several years have been absolutely insane - one "dumpster fire" (to make metaphor) after another. Just to name a few of the more dramatic ones; I've been hospitalized a couple of times (luckily, the conditions were all curable - but they don't just send you home with an IV pump, apparently 😉 ), I've had _very_ time consuming construction problems with our house, and my car got totaled thanks to a hapless deer that jumped out of the bushes _right_ in front of me - ya know, the sort of stuff you'd expect to see on a soap opera. Until now, I haven't had _any_ time to work on this; and my equipment has been un-accessable (too much stuff piled on-top/in-front of it, due to the construction).
> 
> I'm finally starting to dig my way out of it though; and do hope to resume development on this.
> 
> Before everything went sideways in my life I had some stuff written, but untested; once I've gotten my gear accessible and can test that I'll release it.

Hi bdurbrow!

Is there any chance to see you on this project again? I hope so ;) I really appreciate your work on this.
